---
title: "QC Dashboard"
output: 
  flexdashboard::flex_dashboard:
    orientation: rows
    vertical_layout: scroll
    theme: bootstrap
editor_options: 
  chunk_output_type: console
params:
  prefix: files.idepth
  nClusters: 3
  GMKey: NULL
---

```{r setup, include=FALSE}
suppressMessages(library(flexdashboard))
suppressMessages(library(tidyverse))
suppressMessages(library(plotly))
suppressMessages(library(ape))
suppressMessages(library(reshape2))
suppressMessages(library(ggtree))
suppressMessages(library(ggmap))



prefix <- params$prefix
set.colors <- c('#1f78b4','#33a02c','#6a3d9a','#a6cee3','#b2df8a','#fb9a99','#e31a1c','#fdbf6f','#ff7f00','#cab2d6')

```


Individuals
===================================== 

Row
-----------------------------------------------------------------------
### Individuals
```{r}
num.samples <- nrow(read.table(paste0(prefix, ".idepth"), header = T))
valueBox(num.samples, icon = "fa-check-double")
```

### SNPs

```{r}
nums.nps <- nrow(read.table(paste0(prefix, ".bim"), header = T))

valueBox(nums.nps, icon = "fa-dna")
```

### Mean depth

```{r}
mean.depth <- round(mean(read.table(paste0(prefix, ".idepth"), header = T)$MEAN_DEPTH), 1)
valueBox(mean.depth, icon = "fa-align-center")
```

Column {.sidebar}
-----------------------------------------------------------------------
### About

This report is intended to act as a first-pass exploratory analysis of the variants in your dataset. Looking for points that are e.g. extreme outliers in the PCA plot and the depth/missingness plot are a good way to identify potentially problematic (or interesting!) samples. Several basic analysis are included, many of which are colored by the result of a k-means clustering on the first two principal components. A map is possible to generate if you place a tab-separated file with sample ID, longitude, and latitude (decimal degrees) with the extension `.coords` in the directory the contains all of the QC output. By default in this pipeline, these are in the subdirectory `06_QC`. By default, three populations are assumed for k-means clustering, which may be a sensible starting point for visual inspection of your results, but this value can be changed in the config file. 

Row 
-----------------------------------------------------------------------

```{r}
#Column {.sidebar}
#-----------------------------------------------------------------------
#shiny sider bar
#sliderInput('clusters', 'Cluster count', 3,
#              min = 1, max = 6)

input <- list()
input$clusters <- params$nClusters
input$GMKey <- params$GMKey

```

### Genomic PCA

```{r, fig.width = 5, fig.height = 5}
# PCA plot --------------------------------------------------------------
pca.path <- paste0(prefix, ".eigenvec")
#this makes it reasonably robust to running with plink 1.9 or plink 2.0
tmp.head <- sub("#", "", readLines(pca.path))
df.pca <- read.table(text = tmp.head, header = TRUE)
df.val <- read.table(gsub("vec","val", pca.path), header = FALSE)
df.val$prop <- (df.val$V1 / (sum(df.val$V1))) * 100

#add depth
depth.path <- paste0(prefix, ".idepth")
df.depth <- read.table(depth.path, header = T)
df.depth <- df.depth %>%  mutate_if(is.numeric, round, digits = 2)

df.pca <- left_join(df.pca, df.depth, by = c("IID" = "INDV"))

set.seed(42)
k <- kmeans(df.pca[,c("PC1","PC2")], input$clusters, nstart = 25, iter.max = 100)
df.pca$cluster <- paste("cluster",k$cluster, sep = "_")


##PCA plot
(ggplot(data = df.pca, aes(x = PC1, y = PC2, text = IID, fill = cluster), color = "grey90") + 
  geom_point(alpha = 0.9, shape = 21, size = 3) + 
  theme_bw() +
  theme(text = element_text(size = 14),
        legend.position = "none") +
  xlab(paste0("PC1", ": ", round(df.val[1,2],1),"% variance")) +
  ylab(paste0("PC2", ": ", round(df.val[2,2],1),"% variance")) +
  scale_fill_manual(values = set.colors)) %>% 
  ggplotly(tooltip=c("text","x","y")) 

```

### SNP Depth

```{r, fig.width = 5, fig.height = 5}
# depth plot --------------------------------------------------------------
#depth is read in the PCA chunk
imiss.path <- paste0(prefix, ".imiss")
df.imiss <- read.table(imiss.path, header =T)

df.depth.miss <- inner_join(df.depth, df.imiss, by = "INDV")
df.depth.miss <- df.depth.miss %>% 
 mutate_if(is.numeric, round, digits = 5)

#replace df.pca with clusters() for shiny
#r.missing <- reactive(
  df.depth.miss <- left_join(df.depth.miss, df.pca[,c("IID","cluster")], by = c("INDV" = "IID"))
#)

#replace df.depth.miis with r.missing for shiny
#renderPlotly(
  #need to wrap ggplot in () to pipe to plotly
  (ggplot(data = df.depth.miss, aes(text = INDV)) +
    geom_point(aes(x = MEAN_DEPTH, y = F_MISS, fill = cluster), 
               size = 3, color = "black") +
    theme_bw() +
    theme(legend.position = "top") + 
    scale_fill_manual(values = set.colors)+ 
    labs(x = "SNP Depth", y = "Missingness")) %>% 
    ggplotly(tooltip=c("text","x","y"))
#)

```

Row 
-----------------------------------------------------------------------

### Tree 

```{r, fig.height=12, fig.width=12}

dist.id <- read.table(paste0(prefix, ".dist.id"))["V2"]
dist.id <- left_join(dist.id, df.pca[,c("IID","cluster")], by = c("V2" = "IID"))

df.dist <- read.table(paste0(prefix, ".dist"))

mat.dist <- as.dist(df.dist)

nj.dist <- ape::nj(mat.dist)
nj.dist$tip.label <- dist.id$V2

mycol <- set.colors[factor(dist.id$cluster)]

#a messy ape plot
#plot(nj.dist, tip.color = mycol, type = "unrooted", use.edge.length	= F)

# ggtree
p.ggtree <- ggtree(nj.dist, layout="daylight", branch.length = "none")

## with ggtree labeling scheme
# p.ggtreelabs <- p.ggtree %<+% dist.id + 
#   geom_tiplab(pch = 5, aes(col = cluster)) +
#   theme(legend.position = "none") +
#   scale_color_manual(values = set.colors) 

## this tree doesnt work well because the edges get cut off
## due to different axes units in the tree vs text
## ggplot2::xlim(-500, 500) + ylim(-500, 500) sort of helps
# p.ggtreelabs

metat <- p.ggtree$data %>%
  dplyr::inner_join(dist.id, c('label' = 'V2'))

#hacky solution to add text labels
#using the ggtree label geoms is not supported by plotly
p.ggtree.custom <- p.ggtree +
  geom_text(data = metat,
             aes(x = x,
                 y = y,
                 colour = cluster,
                 label = label), size = 3) +
  theme(legend.position = "none") +
  scale_color_manual(values = set.colors) 

p.ggtree.custom

# This interactive code really should work but when rendered with plotly, but in the pipeline, the tree is blank
# this isnt an issue with local install so it may be a ggtree version issue

#ggplotly(p.ggtree.custom, tooltip=c("label", "cluster"))

```

Row
-----------------------------------------------------------------------

### Relatedness matrix

```{r,  fig.height=10, fig.width=6}

df.rel <- read.table(paste0(prefix, ".king"))
colnames(df.rel) <- dist.id$V2
rownames(df.rel) <- dist.id$V2

mat.rel <- as.matrix(df.rel)

#check for -Inf in the matrix
l.inf <- grep("-Inf", mat.rel, value = T)

#clustering will fail if there are -Inf
if(length(l.inf)==0){

  # Run clustering
  rel.dendro <- as.dendrogram(hclust(d = dist(x = mat.rel)))
  sample.order <- order.dendrogram(rel.dendro)
  #we could also plot the dendrogram, but it is potentially confusing
  #with the already plotted NJ tree
  
  #pivot does not seem to work with matrices, so reverting to melt from reshape2
  df.rel.long <- melt(as.matrix(df.rel))
  names(df.rel.long) <- c("Sample1", "Sample2", "relatedness")
  
  matrix.names <- row.names(mat.rel)
  
  df.rel.long$Sample2 <- factor(x = df.rel.long$Sample2,
                                 levels = matrix.names[sample.order], 
                                 ordered = TRUE)
  
  df.rel.long$Sample1 <- factor(x = df.rel.long$Sample1,
                                 levels = matrix.names[sample.order], 
                                 ordered = TRUE)
  
  pl.1 <- (ggplot(data = df.rel.long, aes(x=Sample1, y=Sample2, fill=relatedness)) + 
    geom_tile() +
    theme(
      axis.text.x = element_text(angle = 90),
      axis.title = element_blank(),
      axis.text = element_text(size = 5),
      legend.position = "top"
    ) +
     scale_fill_gradient2(low = "blue", high = "red", mid = "white",
                          midpoint = mean(df.rel.long$relatedness),  
                          name="relatedness")) %>% 
    ggplotly(tooltip=c("relatedness","x","y"))
  
  pl.1

}else{
  print("too little data (SNPs or Individuals) to cluster relatedness matrix")
}

```

Row {data-width=900, data-height=900}
-----------------------------------------------------------------------

### Map

```{r, fig.width = 10, fig.height = 10}

# this code is highly specific to the CCGP as it generates a California map

g <- list(
  scope = 'north america',
  showland = TRUE,
  landcolor = toRGB("grey83"),
  subunitcolor = toRGB("white"),
  countrycolor = toRGB("white"),
  showlakes = TRUE,
  lakecolor = toRGB("white"),
  showsubunits = TRUE,
  showcountries = TRUE,
  resolution = 50,
  projection = list(
    type = 'conic conformal',
    rotation = list(lon = -100)
  ),
  lonaxis = list(
    showgrid = TRUE,
    gridwidth = 0.5,
    range = c(-125, -111),
    dtick = 5
  ),
  lataxis = list(
    showgrid = TRUE,
    gridwidth = 0.5,
    range = c(31, 43),
    dtick = 5
  )
)

if(file.exists(paste0(prefix, ".coords.txt"))){
  if(input$GMKey!=999){
    df.coords <- read.table(paste0(prefix, ".coords.txt"))
    names(df.coords) <- c("sample.ID","long","lat")
    
    df.coords <- left_join(df.pca, df.coords, by = c("IID" = "sample.ID"))
      
    df.coords$color <- set.colors[factor(dist.id$cluster)]
    
    mycol <- set.colors[factor(df.coords$cluster)]
    
    #jitter points
    df.coords$Longitude <- jitter(as.numeric(df.coords$long), 8)
    df.coords$Latitude <- jitter(as.numeric(df.coords$lat), 8)

    register_google(key = input$GMKey)

    us <- c(left = -125, bottom = 25.75, right = -67, top = 49)

    cal.map <- get_googlemap("California", zoom = 6, maptype = "terrain", scale = 4, size = c(640, 640))

    terrain.map <- cal.map %>% 
      ggmap() + 
        geom_point(data = df.coords, aes(x = Longitude, y = Latitude,
                                        color = cluster, size = 2.5, label = IID)) +
        theme_bw() + 
        labs(x = NULL, y = NULL) +
      scale_color_manual(values = set.colors) +
      theme(legend.position = "none")

    ggplotly(terrain.map, tooltip=c("IID","Longitude", "Latitude"))
  
  }else{
    df.coords <- read.table(paste0(prefix, ".coords.txt"))
    names(df.coords) <- c("sample.ID","long","lat")
    
    df.coords <- left_join(df.pca, df.coords, by = c("IID" = "sample.ID"))
      
    df.coords$color <- set.colors[factor(dist.id$cluster)]
    
    mycol <- set.colors[factor(df.coords$cluster)]
    
    df.coords$long.jit <- jitter(as.numeric(df.coords$long), 8)
    df.coords$lat.jit <- jitter(as.numeric(df.coords$lat), 8)
    
    plot_geo(df.coords, lat = ~lat.jit, lon = ~long.jit) %>% 
      layout(legend = list(orientation = 'h'), geo = g) %>% 
      add_markers(
        text = ~paste(IID, cluster, PC1, PC2, sep = "<br />"),
        color = ~cluster, symbol = I("circle"), size = I(60), 
        hoverinfo = "text", colors = set.colors[1:input$clusters],
      ) %>% 
     colorbar(title = "cluster")   
  }
 
}else{
  print("a map will appear here if you includes a .coords file ")
}

```

Row
-----------------------------------------------------------------------

### Admixture

Important to note that the colors in the admixture plot are not neccessarily the same groupings as those in the other plots, as K was set independently of the k-means clustering on the PCA. 


```{r, width = 10, height = 5}
  # admixture ---------------------------------------------------------------

k2 <- paste0(prefix, ".2.Q")
k3 <- paste0(prefix, ".3.Q")
samps <- paste0(prefix, ".fam")

x <- read.table(k2, header = F)

struct_files <- c(k2,k3)
cat_admx <- do.call("rbind",lapply(struct_files,
                               FUN=function(files){
                                 x <- read.table(files, header = F)
                                 names(x) <- gsub("V", "pop", names(x)) #rename ancestral pops
                                 x.samps <- read.table(samps) %>% select(V2) #get sample names from .fam file
                                 x$sampleID <- x.samps$V2 #add sample name to df
                                 x$k <- gsub(".Q","",substr(files, nchar(files)-3+1, nchar(files)))
                                 x.long <- x %>% #pivot longer 
                                   pivot_longer(names_to = "popGroup", values_to = "prob", cols = -c(sampleID, k))
                                 x.long
                               }))

cat_admx.wide <- cat_admx %>% 
  filter(k == 2) %>% 
  select(sampleID, k, popGroup, prob) %>% 
  pivot_wider(names_from = "popGroup", values_from = "prob") %>% 
  arrange(pop1, pop2) %>% 
  mutate(sample_order = 1:n()) %>% 
  ungroup()
sample.order <- cat_admx.wide %>% select(sampleID, sample_order)
cat_admx <- left_join(cat_admx, sample.order, by = c("sampleID"))


##############################

p.k23 <- cat_admx %>% 
  filter(k == 2 | k == 3) %>% 
  ggplot(aes(x = fct_reorder(sampleID, sample_order), y = prob, fill = factor(popGroup), text = sampleID)) +
  geom_col(aes(fill = factor(popGroup)), size = 0.1) +
  facet_grid(rows = vars(k), switch = "x", scales = "free", space = "free") +
  theme_minimal() +
  labs(x = NULL, y = "Ancestry") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_discrete(expand = expansion(add = 1)) +
  theme(
    panel.spacing.x = unit(0.0, "lines"),
    axis.text.x = element_text(angle = 90, size = 6),
    panel.grid = element_blank(),
    strip.text.x = element_text(angle = 90, size = 6),
    legend.position = "none"
  ) +
  #scale_fill_manual(name = "grp",values = c("#1b9e77","#d95f02","#7570b3"), guide = F) +xlab(NULL)  + theme(legend.position = "none")
  scale_fill_manual(name = "grp",values = c("#fb8072","#80b1d3","#8dd3c7"), guide = F) +xlab(NULL)  + theme(legend.position = "none")


#make admix interactive
pl.admix <- ggplotly(p.k23, tooltip=c("text", "y"))
pl.admix
```
